core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1555 @@
+              s"in any live log dir on broker ${replica.brokerId()}. There is likely an " +
+              s"offline log directory on the broker.")
+          case MissingReplicaMoveState(targetLogDir) =>
+            bld.append(s"Partition ${replica.topic()}-${replica.partition()} cannot be found " +
+              s"in any live log directory on broker ${replica.brokerId()}.")
+          case ActiveMoveState(currentLogDir, targetLogDir, futureLogDir) =>
+            if (targetLogDir.equals(futureLogDir)) {
+              bld.append(s"Reassignment of replica ${replica} is still in progress.")
+            } else {
+              bld.append(s"Partition ${replica.topic()}-${replica.partition()} on broker " +
+                s"${replica.brokerId()} is being moved to log dir ${futureLogDir} " +
+                s"instead of ${targetLogDir}.")
+            }
+          case CancelledMoveState(currentLogDir, targetLogDir) =>
+            bld.append(s"Partition ${replica.topic()}-${replica.partition()} on broker " +
+              s"${replica.brokerId()} is not being moved from log dir ${currentLogDir} to " +
+              s"${targetLogDir}.")
+          case CompletedMoveState(targetLogDir) =>
+            bld.append(s"Reassignment of replica ${replica} completed successfully.")
+        }
+    }
+    bld.mkString(System.lineSeparator())
   }
 
-  def generateAssignment(zkClient: KafkaZkClient, brokerListToReassign: Seq[Int], topicsToMoveJsonString: String, disableRackAware: Boolean): (Map[TopicPartition, Seq[Int]], Map[TopicPartition, Seq[Int]]) = {
-    val topicsToReassign = parseTopicsData(topicsToMoveJsonString)
-    val duplicateTopicsToReassign = CoreUtils.duplicates(topicsToReassign)
-    if (duplicateTopicsToReassign.nonEmpty)
-      throw new AdminCommandFailedException("List of topics to reassign contains duplicate entries: %s".format(duplicateTopicsToReassign.mkString(",")))
-    val currentAssignment = zkClient.getReplicaAssignmentForTopics(topicsToReassign.toSet)
+  /**
+   * Clear all topic-level and broker-level throttles.
+   *
+   * @param adminClient     The AdminClient to use.
+   * @param topics                The topics to clear the throttles for.
+   */
+  def clearAllThrottles(adminClient: Admin, topics: Set[String]): Unit = {
+    println("Clearing broker-level throttles....")
+    clearBrokerLevelThrottles(adminClient)
+    println("Clearing topic-level throttles....")
+    clearTopicLevelThrottles(adminClient, topics)
+  }
 
-    val groupedByTopic = currentAssignment.groupBy { case (tp, _) => tp.topic }
-    val rackAwareMode = if (disableRackAware) RackAwareMode.Disabled else RackAwareMode.Enforced
+  /**
+   * Clear all throttles which have been set at the broker level.
+   *
+   * @param adminClient     The AdminClient to use.
+   */
+  def clearBrokerLevelThrottles(adminClient: Admin): Unit = {
+    val allBrokerIds = adminClient.describeCluster().nodes().get().asScala.map(_.id())
```

Comment:
Yes, we do miss brokers that are currently offline.  This is an existing design flaw with how the reassignment tool works.  I think we can improve it somewhat in this change by at least including the brokers that we know about from the reassignment JSON file.

We can't fix it completely, though, since some brokers might have had quotas set, but not be in the JSON file.  If those brokers are down, they will be missed.  For example, if a broker appeared only as a source broker, it won't be in the JSON file.